boundedint/int256.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2020 The Swarm Authors
+// This file is part of the Swarm library.
+//
+// The Swarm library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+package boundedint
+
+import (
+	"fmt"
+	"io"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+// Int256 represents an signed integer of 256 bits
+type Int256 struct {
+	value big.Int
+}
+
+var minInt256 = new(big.Int).Mul(big.NewInt(-1), new(big.Int).Exp(big.NewInt(2), big.NewInt(255), nil)) // -(2^255)
+var maxInt256 = new(big.Int).Sub(new(big.Int).Exp(big.NewInt(2), big.NewInt(255), nil), big.NewInt(1))  // 2^255 - 1
+
+// NewInt256 creates a Int256 struct with an initial underlying value of zero
+func NewInt256() *Int256 {
+	u := new(Int256)
+	u.value = *new(big.Int).Set(big.NewInt(0))
+	return u
+}
+
+// Value returns the underlying private value for a Int256 struct
+func (u *Int256) Value() big.Int {
+	return u.value
+}
+
+// Set assigns the underlying value of the given Int256 param to u, and returns the modified receiver struct
+// returns an error when the result falls outside of the unsigned 256-bit integer range
+func (u *Int256) Set(value big.Int) (*Int256, error) {
+	if value.Cmp(maxInt256) == 1 {
+		return nil, fmt.Errorf("cannot set Int256 to %v as it overflows max value of %v", value, maxInt256)
+	}
+	if value.Cmp(minInt256) == -1 {
+		return nil, fmt.Errorf("cannot set Int256 to %v as it underflows min value of %v", value, minInt256)
+	}
+	u.value = *new(big.Int).Set(&value)
+	return u, nil
+}
+
+// Copy sets the underlying value of u to a copy of the given Int256 param, and returns the modified receiver struct
+func (u *Int256) Copy(v *Int256) *Int256 {
+	u.value = *new(big.Int).Set(&v.value)
+	return u
+}
+
+// Cmp calls the underlying Cmp method for the big.Int stored in a Int256 struct as its value field
+func (u *Int256) Cmp(v *Int256) int {
+	return u.value.Cmp(&v.value)
+}
+
+// Equals returns true if the two Int256 structs have the same underlying values, false otherwise
+func (u *Int256) Equals(v *Int256) bool {
+	return u.Cmp(v) == 0
+}
+
+// Add sets u to augend + addend and returns u as the sum
+// returns an error when the result falls outside of the signed 256-bit integer range
+func (u *Int256) Add(augend, addend *Int256) (*Int256, error) {
+	sum := new(big.Int).Add(&augend.value, &addend.value)
+	return u.Set(*sum)
```

Comment:
this is already kind of done by the `Add` function in `math/big`

Reply:
not sure what you mean here. could you please elaborate? 

`math/big`'s `Add` does not return any kind of error since it is not expecting an over- or under-flow. this would be covered by the `Set` call instead.

on the other hand, the `Add` call for the underlying `big.Int` value is used in order not to have to re-implement the arithmetic functions, which is good for more than one reason.

Reply:
why can't you use `u.value.Add(...)` directly?

Reply:
`u.value` is a `big.Int`-typed variable, therefore:
- it does not control the over or underflow that needs to be checked for the `Int256` type when calling `Add` from it
- if `Add` is called directly from `value` (`int256.go:78`) it will change the value before making this check 

we _could_ reset it to its previous value if the check is done after calling `Add`, but why would we?